File: src/csharp/9_easy_回文数/S1.cs
/*
 * @lc app=leetcode.cn id=9 lang=csharp
 *
 * [9] 回文数
 */

namespace LeetCode.P9.S1;

// @lc code=start
public class Solution
{
    public bool IsPalindrome(int x)
    {
        if (x < 0) return false;
        if (x < 10) return true;

        // 得到数字位数
        var digit = 0;
        var temp = x;
        while (temp > 0)
        {
            temp = temp / 10;
            digit++;
        }

        for (var i = 1; i <= digit / 2; i++)
        {
            var left = (byte)(x / Math.Pow(10, digit - i) % 10);
            var right = (byte)(x / Math.Pow(10, i - 1) % 10);
            if (left != right) return false;
        }

        return true;
    }
}
// @lc code=end

public static class Program
{
    public static void Main()
    {
        var solution = new Solution();

        Console.WriteLine(solution.IsPalindrome(1000021));
    }
}

File: src/python3/20_easy_有效的括号/S1.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#


# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        counts = [0, 0, 0]
        left = ["(", "[", "{"]
        right = [")", "]", "}"]
        pairs = {"(": ")", "[": "]", "{": "}"}
        stack = []

        for ch in s:
            if ch in left:
                counts[left.index(ch)] += 1
                stack.append(ch)
            else:  # right
                counts[right.index(ch)] -= 1
                if len(stack) == 0:
                    return False
                if pairs[stack[-1]] != ch:
                    return False
                stack.pop()

        return counts == [0, 0, 0]


# @lc code=end
